refactor: 예외 처리 로직 개선

ahyeonkong opened the pull request:
## #️⃣ Issue Number

#92 

## 📝 작업 설명

- Controller의 불필요한 try-catch를 service로 옮기기
- 세션 관련 예외 처리 적용하기
- IOException을 CustomException으로 변경해서 응답 메시지 반환하기
- ErrorCode enum 추가하기
- Controller의 throws 예외 처리 제거

## 📸스크린샷 (선택)   

## 💬 리뷰 요구사항

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [x] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

ahyeonkong committed:
refactor: 프로필 사진 관련 예외 처리 로직 개선

- controller 메서드에서 try-catch 블록 제거
- HttpSessionRequiredException을 throws로 처리
- S3Service의 uploadProfileImage 메서드에서 사용자 조회 실패 시 CustomException 사용
- IOException을 CustomException으로 변환하여 처리

Resolves: #92

ahyeonkong committed:
refactor: ErrorCode를 활용한 예외 처리 개선

- 기존 RuntimeException 및 IllegalArgumentException을 CustomException으로 대체하여 일관된 예외 처리 구현

Resolves: #92

ahyeonkong committed:
refactor: 사진 관련 예외 처리 로직 개선

- controller에서 try-catch 제거
- S3Service의 uploadFile 메서드에서 IOException을 CustomException으로 변환

Resolves: #92

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/controller/PhotoController.java:
```diff
@@ -27,21 +25,15 @@ public PhotoController(AuthService authService, S3Service s3Service) {
-        } catch (IOException e) {
-            return ApiResponse.error(ErrorCode.FILE_UPLOAD_ERROR);
+        List<String> fileUrls = new ArrayList<>();
+        for (MultipartFile file : files) {
```
업로드 관련한 코드도 Service 단으로 이동 시키는 게 좋을 것같아요!

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/controller/ProfileImageController.java:
```diff
@@ -23,27 +21,17 @@ public class ProfileImageController {
-        }
+    public ApiResponse<String> uploadProfileImage(@RequestParam("file") MultipartFile file, HttpServletRequest request) throws HttpSessionRequiredException {
+        String email = authService.getAuthenticatedEmail(request);
+        String fileUrl = s3Service.uploadProfileImage(file, email);
```
Service 서로 간의 참조 없이 코드를 작성하기 위해 Controller 단에 코드를 작성하시게 된 것같아요!
해당 사항은 준배님과도 결정해야할 사항이지 않을까 싶습니다만,

1. 아현님이 작성해주신 코드처럼 Controller에서 필요한 모든 Service를 참조해 사용한다.
2. 하나의 Controller 에는 하나의 Service 만을 사용한다.

2번의 경우 Service 단 내부에 getAuthenticatedEmail 메소드를 이동, 혹은 생성하거나 Util, Service 보다 밑단의 코드를 작성하는 방법이 있을 것같네요!

ahyeonkong replied:
오 함께 논의해볼 사항인 것 같네요!

1번 방법의 경우, 하나의 Controller에서 여러 Service를 재사용하면 필요한 요청을 한 곳에서 처리하기 쉬울 것 같아서 위 방식을 선택했습니다.
찾아보니 2번 방법의 경우, 하나의 Controller에 하나의 Service를 사용하는 방식(SRP)을 활용하네요.

아직은 주입 받은 Service가 많지 않아서 1번 방법을 유지해도 좋지만, 서비스 규모가 커져서 3개 이상 주입 받게 되면 2번 방법을 선택해도 좋겠다는 의견입니다!

JinTaekLim replied:
@JunBe 준배님께선 어떻게 생각하실까요?

JunBe replied:
저는 1번이 조금 더 괜찮은 것 같습니다!

AuthService의 경우 여러 컨트롤러에서 사용 될 수 있을 것 같은데, 여기 있는 메서드를 s3Service로 이동 시키면 다른 서비스에서도 동일한 인증 관련 서비스가 필요한 경우에 해당 메서드에 중복되는 코드가 작성 될 가능성이 있습니다.

그리고 컨트롤러에서 로직을 처리하는게 아니기 때문에 SRP 원칙을 위배한다고 보지 않아도 될 것 같습니다.
컨트롤러에서 각 서비스를 호출해서 어떤 흐름으로 이어지는지 한 눈에 더 잘 파악할 수 있을 것 같아서 1번 방식으로 해도 좋을 것 같습니다!

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/controller/ProfileImageController.java:
```diff
@@ -23,27 +21,17 @@ public class ProfileImageController {
-        }
+    public ApiResponse<Map<String, String>> getProfileImageUrl(HttpServletRequest request) throws HttpSessionRequiredException {
+        String email = authService.getAuthenticatedEmail(request);
+        Map<String, String> urls = s3Service.getProfileImageUrl(email);
```
DTO가 아닌 Map을 사용하신 이유가 있을까요?

ahyeonkong replied:
기존 코드가 Map으로 사용되어 있어, 그대로 사용했습니다!

JinTaekLim replied:
이후 어떤 값이 필요한지 직관적으로 확인할 수 있으니 가능하면 DTO로 사용하는게 좋을 것같아요!
추가적으로 Controller 단에선 예외 처리를 하지 않는게 좋을 것같다는 생각입니다만! ( thorws ) 혹시 어떻게 생각하실까요?

JunBe replied:
Map으로 반환 타입을 결정한 이유가 여러 개의 사진을 받을 때, 응답 값으로 프리사인 url만 나오면 어떤 사진이 어떤 url이 어떤 사진 일지 헷갈릴 것 같아서 {s3 객체 url : 프리사인 url} 형식으로 반환하도록 Map을 이용했었습니다.

DTO를 만든다 하면 그대로 Map 타입을 이용하는 게 좋을까요 List<String> 타입을 사용하여 프리사인 url만 응답하도록 하는 게 좋을까요?

JunBe replied:
저도 예외처리는 서비스 단에서 하는게 좋을 것 같습니다!

ahyeonkong replied:
1. Map 타입
{s3 객체 url : 프리사인 url} 형식으로 데이터를 반환한다면 추후 클라이언트가 쉽게 이해할 수 있을 것 같다는 의견입니다!
``` java
{
  "https://s3.amazonaws.com/mybucket/file1.jpg": "https://presigned-url-for-file1",
  "https://s3.amazonaws.com/mybucket/file2.jpg": "https://presigned-url-for-file2"
}
```
 따라서 그대로 Map 타입을 유지하되, DTO 안에 Map을 넣는 형식이 명확한 것 같습니다.

만약 List를 사용하게 되면
``` java
[
  "https://presigned-url-for-file1",
  "https://presigned-url-for-file2"
]
```
다음과 같이 어떤 s3 객체가 프리사인 url을 반환하는지 알기 어렵지 않을까라는 생각이 듭니다.

2. 저도 예외처리는 서비스 단에서 하는 게 깔끔할 것 같습니다!

JinTaekLim replied:
-> Map 은 사용하지 않고 DTO 를 활용해 접근 가능한 URL 하나만 반환하도록 결정

JinTaekLim commented on src/main/java/com/market/saessag/domain/photo/service/S3Service.java:
```diff
@@ -50,7 +49,7 @@ public void init() {
 
         if (accessKey == null || secretKey == null) {
-            throw new IllegalArgumentException("액세스 키, 또는 시크릿 키 환경 변수가 설정되지 않았습니다.");
+            throw new CustomException(ErrorCode.KEY_CREDENTIALS_MISSING);
```
현재는 S3 와 연동이 불가할 경우 서버가 종료되고 있는데, 
만약 S3에 장애가 발생해 복구가 어려운 경우에는 어떻게 하면 좋을까요?

ahyeonkong replied:
AWS 내의 백업 서비스 기능을 활용하면 어떨까 싶습니다

JinTaekLim replied:
혹시 AWS 와 통신(연결)이 되지 않는 상황에서도 백업 서비스 기능이 사용 가능한가요?

ahyeonkong replied:
지금까지 구현한 것으로는 AWS와 통신이 끊긴 상태에서는 백업 서비스를 사용하기 어려울 것 같습니다 😂

만약 백업 서비스를 유지하고 싶다면 AWS Storage Gateway를 사용하는 방법이 있다고는 하지만, 추가적인 설정이 필요할 것 같습니다!

ahyeonkong committed:
fix: 충돌 에러 해결

ahyeonkong committed:
Merge branch 'develop' of https://github.com/saessagMarket/back into refactor/#92-improve-exception-handling

ahyeonkong committed:
refactor: 컨트롤러의 파일 업로드 로직을 PhotoUploadService로 이동

- 새로운 PhotoUploadService를 생성하여 파일 업로드 로직을 처리하도록 구현
- S3Service는 AWS S3와의 상호작용만 담당하도록 역할 분리

ahyeonkong committed:
chore: SessionUtils 추가 전 관련 코드 주석 처리
